In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path=r'/Users/sofiakazeko/Jupiter/Dec-2024 Instacart Basket Analysis'

In [5]:
#3 import the orders_products_combined
df_orders_products_combined= pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [9]:
# Check the shape of the imported dataframe 
df_orders_products_combined.shape

(32434489, 12)

In [8]:
df_orders_products_combined.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_time  days_since_prior_order  new_customer  product_id  \
0           8                     NaN          True         196   
1           8                     NaN          True       14084   
2           8                     NaN          True       12427   
3           8                     NaN          True       26088   
4           8                     NaN          True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [10]:
#import df_products
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [11]:
df_prods.shape

(49672, 6)

In [12]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

# Determine a suitable way to combine the dataframes

When we have two DataFrames with a common column like "product_id" and assume the values match, an inner merge is a good approach. It combines rows where "product_id" matches in both DataFrames and leaves out non-matching rows. If our assumption is correct, the result will include all relevant rows. If not, the inner merge helps us identify mismatches by excluding unmatched data.

In [14]:
#merging on product id
df_ords_prods_merge = df_orders_products_combined.merge(df_prods, on = 'product_id', indicator = 'merge_2')

In [15]:
df_ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             0   2539329        1             1                   2   
2             0   2539329        1             1                   2   
3             0   2539329        1             1                   2   
4             0   2539329        1             1                   2   

   order_time  days_since_prior_order  new_customer  product_id  \
0           8                     NaN          True         196   
1           8                     NaN          True       14084   
2           8                     NaN          True       12427   
3           8                     NaN          True       26088   
4           8                     NaN          True       26405   

   add_to_cart_order  reordered _merge  Unnamed: 0_y  \
0                  1          0   both           195   
1                  2          0   both         14084   
2                  3          0   both         12427   
3                  4          0   both         26089   
4                  5          0   both         26406   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_2  
0    both  
1    both  
2    both  
3    both  
4    both

# Confirm the results of the merge using the merge flag

In [16]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [17]:
df_ords_prods_merge.shape

(32404859, 18)

Before merge: 32,434,489 rows
After merge: 32,404,859 rows
This means that 29,630 rows (32,434,489 - 32,404,859) did not have matching values in the merge key and were excluded.

In [18]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_ords_prods_merge.pkl'))